diff --git a/src/llama-context.cpp b/src/llama-context.cpp
index 7da4dd1fd..bd68f8137 100644
--- a/src/llama-context.cpp
+++ b/src/llama-context.cpp
@@ -510,7 +510,12 @@ void llama_context::sched_reserve(uint32_t n_tokens, uint32_t n_seqs, uint32_t n
                 }
                 ggml_backend_dev_t device_gdn = ggml_backend_get_device(ggml_backend_sched_get_tensor_backend(sched.get(), n));
 
-                GGML_ASSERT(strncmp(n->name, LLAMA_TENSOR_NAME_FGDN_AR "-", prefix_len) == 0);
+                if (strncmp(n->name, LLAMA_TENSOR_NAME_FGDN_AR "-", prefix_len) != 0) {
+                    LLAMA_LOG_WARN("%s: unexpected fused Gated Delta Net (autoregressive) tensor '%s', set to disabled\n", __func__, n->name);
+                    gdn_device_mismatch = true;
+                    break;
+                }
+
                 const int il = std::stoi(n->name + prefix_len);
                 ggml_backend_dev_t device_kv = model.dev_layer(il);
                 if (device_gdn != device_kv) {
@@ -551,7 +556,12 @@ void llama_context::sched_reserve(uint32_t n_tokens, uint32_t n_seqs, uint32_t n
                 }
                 ggml_backend_dev_t device_gdn = ggml_backend_get_device(ggml_backend_sched_get_tensor_backend(sched.get(), n));
 
-                GGML_ASSERT(strncmp(n->name, LLAMA_TENSOR_NAME_FGDN_CH "-", prefix_len) == 0);
+                if (strncmp(n->name, LLAMA_TENSOR_NAME_FGDN_CH "-", prefix_len) != 0) {
+                    LLAMA_LOG_WARN("%s: unexpected fused Gated Delta Net (chunked) tensor '%s', set to disabled\n", __func__, n->name);
+                    gdn_device_mismatch = true;
+                    break;
+                }
+
                 const int il = std::stoi(n->name + prefix_len);
                 ggml_backend_dev_t device_kv = model.dev_layer(il);
                 if (device_gdn != device_kv) {
